chia/types/eligible_coin_spends.py

```diff
@@ -97,12 +222,138 @@ def get_deduplication_info(
+
+        Raises:
+            If a fast forward cannot proceed, to prevent potential double spends
+        """
+        new_coin_spends = []
+        ff_bundle_coin_spends = {}
+        for coin_id, spend_data in mempool_item.bundle_coin_spends.items():
+            if not spend_data.eligible_for_fast_forward:
+                # Nothing to do for this spend, moving on
+                new_coin_spends.append(spend_data.coin_spend)
+                continue
+            # See if we processed a fast forward item with this puzzle hash before
+            unspent_lineage_info = self.fast_forward_spends.get(spend_data.coin_spend.coin.puzzle_hash)
+            if unspent_lineage_info is None:
+                # We didn't, so let's lookup the most recent version from the DB
+                unspent_lineage_info = await get_unspent_lineage_info_for_puzzle_hash(
+                    spend_data.coin_spend.coin.puzzle_hash
+                )
+                if unspent_lineage_info is None:
+                    raise ValueError("Cannot proceed with singleton spend fast forward.")
+                # See if we're the most recent version
+                if unspent_lineage_info.coin_id == coin_id:
+                    # We are, so we don't need to fast forward, we just need to
+                    # set the next version from our additions to chain ff spends
+                    set_next_singleton_version(
+                        current_singleton=spend_data.coin_spend.coin,
+                        singleton_additions=spend_data.additions,
+                        fast_forward_spends=self.fast_forward_spends,
+                    )
+                    # Nothing more to do for this spend, moving on
+                    new_coin_spends.append(spend_data.coin_spend)
+                    continue
+                # We're not the most recent version, so let's fast forward
+                new_coin_spend, patched_additions = perform_the_fast_forward(
+                    unspent_lineage_info=unspent_lineage_info,
+                    spend_data=spend_data,
+                    fast_forward_spends=self.fast_forward_spends,
+                )
+                # Mark this coin for a coin spend data update
+                ff_bundle_coin_spends[coin_id] = BundleCoinSpend(
+                    coin_spend=new_coin_spend,
+                    eligible_for_dedup=spend_data.eligible_for_dedup,
+                    eligible_for_fast_forward=spend_data.eligible_for_fast_forward,
+                    additions=patched_additions,
+                    cost=spend_data.cost,
+                )
+                # Update the list of coins spends that will make the new fast
+                # forward spend bundle
+                new_coin_spends.append(new_coin_spend)
+                # We're done here, moving on
+                continue
+            # We processed this puzzle hash before, so build on that
+            # See first if we're the most recent version
```

Comment:
I'd prefer using more concrete language. when you say "processed" you mean "added a spend of it" to the block we're building, right?

How would it be possible for our coin to be the most recent version?
I don't think you can submit a transaction to the mempool that spends the output of another transaction already in the pool.